ls1intum/Artemis
# `Programming exercises`: Always give students read permissions for their repository

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
Closes #5876
Closes #5877

### Description
Ensures that students always get read permissions for their repositories, even if only the online editor is enabled. This enables the review of the code for exam programming exercises. For regular exercises this is enabled as well, both for consistency and as I can see no disadvantage.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 2 Programming Exercises
  - one that has the offline IDE disabled
  - one for which the offline IDE is enabled

1. Ensure that a due date is set for the programming exercise.
2. As student: participate in the exercise.
3. If offline IDE disabled: check that you cannot push to the repo to any branch.
4. After the due date: Check that you can no longer push to the repository.
5. For a fresh exercise repeat the steps, but as instructor change the settings for the offline IDE while the exercise is running. The permissions should be removed from the student repo when disabling the offline IDE, and added back when allowing it again.


#### Exam Mode Testing



Prerequisites:
- 1 Instructor
- 2 Students
- 1 (Test) Exam with two Programming Exercises

1. Configure one programming exercise to only allow the online editor, one that also/alternatively allows the offline IDE.
6. Participate in the exam as a student. You should be able to submit to the programming exercise.
7. After the exam, on the summary page: There is a link to the repository for the exercise. That link should work so that you can view the code in the version control system, but not commit further changes.

### Review Progress






#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage




| Class/File | Branch | Line |
|------------|-------:|-----:|
| ExerciseService.java | 85% | 77% |
| AbstractVersionControlService.java | 100% of changed code | 100% of changed code |
| BitbucketService.java | 100% of changed code | 100% of changed code |
| GitLabService.java | 50% | 82% |
